### PR TITLE
[MU4] Added buffered paint provider

### PR DESCRIPTION
**Pull request by igorkorsukov**
see also https://github.com/musescore/MuseScore/pull/7614

### CODE REVIEW

File: src/libmscore/draw/bufferedpaintprovider.h [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+#include <QPen>
+#include <QBrush>
+#include <QFont>
+
+#include "ipaintprovider.h"
+#include "drawtypes.h"
+
+namespace mu::draw {
+struct DrawBuffer
```
Comment (RomanPudashkin) [inline src/libmscore/draw/bufferedpaintprovider.h:32]: I think we can move this struct to a private part of BufferedPaintProvider

File: src/libmscore/draw/bufferedpaintprovider.h [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+    void drawText(const QPointF& p, const QString& s) override;
+    void drawText(const QRectF& r, int flags, const QString& text) override;
+
+    void drawGlyphRun(const QPointF& position, const QGlyphRun& glyphRun) override;
+
+    void fillRect(const QRectF& r, const QBrush& brush) override;
+
+    void drawPixmap(const QPointF& p, const QPixmap& pm) override;
+    void drawTiledPixmap(const QRectF& rect, const QPixmap& pm, const QPointF& offset = QPointF()) override;
```
Comment (RomanPudashkin) [inline src/libmscore/draw/bufferedpaintprovider.h:140]: We use different namings in different methods for rects (r vs rect). I think it is better to use "rect" in all places

File: src/libmscore/draw/painter.cpp [hunk 3]
```diff
@@ -163,91 +227,158 @@ QRect Painter::viewport() const
...
 
-void Painter::drawArc(const QRectF& rect, int a, int alen)
+void Painter::drawArc(const QRectF& r, int a, int alen)
 {
-    m_provider->drawArc(rect, a, alen);
+    QRectF rect = r.normalized();
+
+    QPainterPath path;
+    path.arcMoveTo(rect, a / 16.0);
+    path.arcTo(rect, a / 16.0, alen / 16.0);
```
Comment (RomanPudashkin) [inline src/libmscore/draw/painter.cpp:322]: Can you explain please what this number means?

File: src/libmscore/draw/bufferedpaintprovider.cpp [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    for (int i = 1; i < pointCount; ++i) {
+        QPointF p1 = points[i - 1];
+        QPointF p2 = points[i];
+        d.lines.push_back(QLineF(p1, p2));
+    }
+}
+
+void BufferedPaintProvider::drawPolygon(const QPointF* points, int pointCount, Qt::FillRule fillRule)
+{
+    QVector<QPointF> vec(pointCount);
```
Comment (RomanPudashkin) [inline src/libmscore/draw/bufferedpaintprovider.cpp:242]: We can use QPolygonF instead of this

**Review by RomanPudashkin — APPROVED** [review]

File: src/libmscore/draw/bufferedpaintprovider.h (hunk 2 above)
Comment (igorkorsukov) [reply src/libmscore/draw/bufferedpaintprovider.h:140]: fixed

File: src/libmscore/draw/painter.cpp (hunk 3 above)
Comment (igorkorsukov) [reply src/libmscore/draw/painter.cpp:324]: This is copied from the QPainter source code.
Added a comment about it.

File: src/libmscore/draw/bufferedpaintprovider.cpp (hunk 4 above)
Comment (igorkorsukov) [reply src/libmscore/draw/bufferedpaintprovider.cpp]: fixed

**Pull request merged**